crates/sparrow-runtime/src/prepare/execute_input_stream.rs

```diff
@@ -0,0 +1,468 @@
+            // could wrap the underlying reader with "late data handling" before this step.
+
+            // 1. Drop all "late data" from the batch
+            let time_column = unfiltered_batch
+                .column_by_name(&config.time_column_name)
+                .expect("time column");
+            let time_column = arrow::compute::cast(time_column, &TimestampNanosecondType::DATA_TYPE)
+                .into_report()
+                .change_context(Error::PreparingColumn)?;
+            let time_column: &TimestampNanosecondArray = downcast_primitive_array(time_column.as_ref())
+                .into_report()
+                .change_context(Error::PreparingColumn)?;
+
+            if input_buffer.watermark < 0 {
+                debug_assert!(
+                    time_column.value(0) - bounded_lateness > 0,
+                    "invalid time; below 0",
+                );
+                // If there's no watermark yet, initialize it to the first event time - the bounded lateness
+                input_buffer.watermark = time_column.value(0) - bounded_lateness
+            };
+
+            // Find which indices to take from the batch (the non-late data)
+            let take_indices = time_column
+                .iter()
+                .enumerate()
+                .filter_map(|(index, time)| {
+                    let time = time.expect("valid time");
+                    if time >= input_buffer.watermark {
+                        input_buffer.watermark = std::cmp::max(input_buffer.watermark, time - bounded_lateness);
+                        Some(index as u64)
+                    } else {
+                        // Late data - drop it
+                        None
+                    }
+                })
+                .collect::<Vec<u64>>();
+            let take_indices: PrimitiveArray<UInt64Type> = PrimitiveArray::from_iter_values(take_indices);
+
+            // Drop all rows with late data from each column
+            let filtered_columns = unfiltered_batch
+                .columns()
+                .iter()
+                .map(|column| arrow::compute::take(column, &take_indices, None))
+                .try_collect()
+                .into_report()
+                .change_context(Error::PreparingColumn)?;
+            let record_batch = RecordBatch::try_new(unfiltered_batch.schema(), filtered_columns)
+                .into_report()
+                .change_context(Error::PreparingColumn)?;
+
+            // 2. Slicing may reduce the number of entities to operate and sort on.
+            let record_batch = slice_preparer.slice_batch(record_batch)?;
+
+            // 3. Prepare each of the columns by getting the column behavior result
+            let mut prepared_columns: Vec<ArrayRef> = Vec::new();
+            for c in columns.iter_mut() {
+                let result = c.get_result(&record_batch).await?;
+                update_key_inverse(c, &record_batch, &result, key_hash_inverse.clone()).await?;
```

Comment:
This seems wrong -- isn't this updating the key hash inverse *for every column*?

I think we'd want something like the following, but outside of the loop:

```
let mut prepared_columns: Vec<ArrayRef> = Vec::new();
let key_column = prepared_columns.get_index(key_column);
let key_hash_column= prepared_columns.get_index(key_hash_column);
update_key_hash_inverse(...)
```

Reply:
How it currently is is mimicking what was there - we're matching on the column behavior and updating when we see the `ColumnBehavior::EntityKey`. I suppose we could change it, since the only metadata that's added is from the key

Reply:
I think that would feel a lot more straightforward. Even if we don't use the name/index to look it up, it would be clearer if that happened outside the loop. This makes it really easy to read this as updating on every column.

Reply:
Cool, yeah. Moved this out